site/content/blog/multi-env-deployment-ytt-kapp.md

```diff
@@ -0,0 +1,331 @@
+^          sample-app-staging    Deployment  -       -    create  -       reconcile  -   -  
+^          sample-app-staging    Service     -       -    create  -       reconcile  -   -  
+
+Op:      3 create, 0 delete, 0 update, 0 noop, 0 exists
+Wait to: 3 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: 
+```
+
+Here [kapp](https://carvel.dev/kapp/) is showing calculated changes between configuration provided and live cluster state and asking for confirmation before actually applying the change.
+Similarly, for deploying application on prod environment we will be passing `values-prod.yaml`.
+
+### Overlays
+When user would like to configure fields beyond what the original author has exposed as data values, they should turn to Overlays. Here we look into a way to specify locations within configuration and either add to, remove from, or replace within that existing configuration.
+
+```yaml
+#! add-namespace.yaml
+#@ load("@ytt:overlay", "overlay")
+
+#@overlay/match by=overlay.all, expects="1+"
+---
+metadata:
+  #@overlay/match missing_ok=True
+  namespace: my-namespace
+```
+
+Let's now consider a usecase where we want to patch the namespace for all of the resources we have and we don't want to edit all the documents to add the `namespace` field. This can easily be achieved by something called as [Overlays](https://carvel.dev/ytt/docs/v0.40.0/ytt-overlays/) in ytt. In `add-namespace.yaml` file we have defined an overlay to add the `namespace` field inside `metadata` for all the resources. Here ` #@overlay/match missing_ok=True` means that let's add the field even if it not exists and if it exists let's change the namespace to `my-namespace`.
+
+On applying overlay, final manifest will look like - 
+
+```bash
+$ ytt -f app.yaml -f schema.yaml -f values-prod.yaml -f add-namespace.yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: sample-app-prod
+  labels:
+    app.kubernetes.io/name: sample-app-prod
+  namespace: my-namespace
+spec:
+  selector:
+    app: sample-app-prod
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 8080
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: sample-app-prod
+  labels:
+    app.kubernetes.io/name: sample-app-prod
+  namespace: my-namespace
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: sample-app-prod
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: sample-app-prod
+    spec:
+      containers:
+      - name: sample-app
+        image: nginx:1.14.2
+        imagePullPolicy: IfNotPresent
+        ports:
+        - containerPort: 8080
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: application-settings
+  namespace: my-namespace
+data:
+  app_mode: prod
+  certificates: /etc/ssl/prod
+  db_user: prod-user
+  db_password: prod-password
+```
+
+**Try this** out in our [ytt interactive playground](https://carvel.dev/ytt/#gist:https://gist.github.com/pivotaljohn/495a132fdce5a6889fd5ccf0b48b098f). 
```

Comment:
Love the invitation to play. :) 

Should we add at least a note on what to do to generate the production version? or to see the default version?